tests/suites/test_suite_ctr_drbg.function

```diff
@@ -54,83 +63,107 @@ exit:
     add1_len = unhexify( add1, add1_string );
     add2_len = unhexify( add2, add2_string );
     result_len = unhexify( result, result_string );
 
-    test_offset_idx = 0;
-    test_max_idx = entropy_len;
     /* CTR_DRBG_Instantiate(entropy[:entropy_len], nonce, perso, <ignored>)
-     * where nonce||perso = add_init[add_init_len] */
-    TEST_ASSERT( mbedtls_ctr_drbg_seed_entropy_len( &ctx, mbedtls_test_entropy_func, entropy, add_init, add_init_len, entropy_len ) == 0 );
+     * where nonce||perso = nonce[nonce_len] */
+    TEST_ASSERT( mbedtls_ctr_drbg_seed_entropy_len(
+                     &ctx,
+                     mbedtls_test_entropy_func, entropy,
+                     nonce, nonce_len,
+                     entropy_chunk_len ) == 0 );
+    if( reseed_mode == RESEED_ALWAYS )
+        mbedtls_ctr_drbg_set_prediction_resistance(
+            &ctx,
+            MBEDTLS_CTR_DRBG_PR_ON );
+
+    if( reseed_mode == RESEED_FIRST )
+    {
+        /* CTR_DRBG_Reseed(entropy[idx:idx+entropy_len],
+         *                 reseed[:reseed_len]) */
+        TEST_ASSERT( mbedtls_ctr_drbg_reseed(
+                         &ctx,
+                         reseed, reseed_len ) == 0 );
+    }
 
     /* CTR_DRBG_Generate(result_len * 8 bits, add1[:add1_len]) -> buf */
-    TEST_ASSERT( mbedtls_ctr_drbg_random_with_add( &ctx, buf, result_len, add1, add1_len ) == 0 );
+    /* Then reseed if prediction resistance is enabled. */
+    TEST_ASSERT( mbedtls_ctr_drbg_random_with_add(
+                     &ctx,
+                     buf, result_len,
+                     add1, add1_len ) == 0 );
+
+
+    if( reseed_mode == RESEED_SECOND )
+    {
+        /* CTR_DRBG_Reseed(entropy[idx:idx+entropy_len],
+         *                 reseed[:reseed_len]) */
+        TEST_ASSERT( mbedtls_ctr_drbg_reseed(
+                         &ctx,
+                         reseed, reseed_len ) == 0 );
+    }
+
     /* CTR_DRBG_Generate(result_len * 8 bits, add2[:add2_len]) -> buf */
-    TEST_ASSERT( mbedtls_ctr_drbg_random_with_add( &ctx, buf, result_len, add2, add2_len ) == 0 );
+    /* Then reseed if prediction resistance is enabled. */
+    TEST_ASSERT( mbedtls_ctr_drbg_random_with_add(
+                     &ctx,
+                     buf, result_len,
+                     add2, add2_len ) == 0 );
     TEST_ASSERT( memcmp( buf, result, result_len ) == 0 );
 
 exit:
     mbedtls_ctr_drbg_free( &ctx );
 }
 /* END_CASE */
+
+/* BEGIN_CASE */
+void ctr_drbg_validate_no_reseed( char *add_init_string, char *entropy_string,
+                                  char *add1_string, char *add2_string,
+                                  char *result_string )
+{
+    test_suite_ctr_drbg_validate( RESEED_NEVER, add_init_string,
+                                  strlen( entropy_string ) / 2, entropy_string,
+                                  "", add1_string, add2_string,
+                                  result_string );
+    goto exit;
```

Comment:
Why do we need these "goto exit;" in all the tests?

Reply:
The root cause is that the function `ctr_drbg_validate` is both used as a test function, and as an auxiliary function with some trivial wrappers. That's not something the test framework supports well. It's a test function so it returns `void`. The trivial wrappers are left with no source of failure, so no natural `goto exit` (which is normally hidden inside a `TEST_ASSERT`). But the test framework adds an `exit` label to test functions if they don't have one already, and many compilers complain about an unused label.

I think it would be best to change `ctr_drbg_validate` into an auxiliary function that returns `int`, and to have four simple wrapper functions (one for each `RESEED_xxx`) that each call `TEST_ASSERT( ctr_drbg_validate( … ) )`. This means creating a wrapper function for `RESEED_FIRST`. I didn't have time to do this on Friday.